src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnection.cs

```diff
@@ -0,0 +1,576 @@
+        {
+            _pipe = namedPipeClient ?? throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+            if (namedPipeClient == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+            if (!namedPipeClient.IsConnected)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+
+            _pipe = namedPipeClient;
+            _closeState = CloseState.Open;
+            _exceptionEventType = TraceEventType.Error;
+            _connectionBufferSize = connectionBufferSize;
+            _atEOFEvent = new ManualResetEventSlim(false);
+        }
+
+        public int ConnectionBufferSize
+        {
+            get
+            {
+                return _connectionBufferSize;
+            }
+        }
+
+        private static byte[] ZeroBuffer
+        {
+            get
+            {
+                return s_zeroBuffer;
+            }
+        }
+
+        public TraceEventType ExceptionEventType
+        {
+            get { return _exceptionEventType; }
+            set { _exceptionEventType = value; }
+        }
+
+        public void Abort()
+        {
+            Abort(null, TransferOperation.Undefined);
+        }
+
+        private void Abort(string timeoutErrorString, TransferOperation transferOperation)
+        {
+            CloseHandle(true, timeoutErrorString, transferOperation);
+        }
+
+        private Exception ConvertPipeException(PipeException pipeException, TransferOperation transferOperation)
+        {
+            return ConvertPipeException(pipeException.Message, pipeException, transferOperation);
+        }
+
+        private Exception ConvertPipeException(string exceptionMessage, PipeException pipeException, TransferOperation transferOperation)
+        {
+            if (_timeoutErrorString != null)
+            {
+                if (transferOperation == _timeoutErrorTransferOperation)
+                {
+                    return new TimeoutException(_timeoutErrorString, pipeException);
+                }
+                else
+                {
+                    return new CommunicationException(_timeoutErrorString, pipeException);
+                }
+            }
+            else if (_aborted)
+            {
+                return new CommunicationObjectAbortedException(exceptionMessage, pipeException);
+            }
+            else
+            {
+                return new CommunicationException(exceptionMessage, pipeException);
+            }
+        }
+
+        public async ValueTask<int> ReadAsync(Memory<byte> buffer, TimeSpan timeout)
+        {
+            ValidateBufferBounds(buffer);
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            var cancellationToken = await timeoutHelper.GetCancellationTokenAsync();
```

Comment:
Why is GetCancellationTokenAsync async? Seems like an odd place to do IO

Reply:
It's not doing IO, it's coalescing CancellationToken's based on deadline. This was a mitigation to reduce load on the TimerManager. As two threads could be requesting a CancellationToken for the same deadline (within a window), the second thread that tries to get an instance will have to wait for the first one to finish initializing it. Think of it like an asynchronous version of Lazy\<T\> where the loser of the race condition doesn't block waiting for the winner to finish.